### PR TITLE
feat:social account initialize with nostr public key

### PR DESCRIPTION
**Pull request by mubarak23**
Social Account Contract #16 
constructor that single storage slot - public_key

### CODE REVIEW

File: onchain/src/social_account.cairo [hunk 1]
```diff
@@ -0,0 +1,20 @@
+#[starknet::contract]
+pub mod SocialAccount {
```
Comment (maciejka) [inline onchain/src/social_account.cairo:2]: Rename to SocialPayAccount

File: onchain/src/social_account.cairo [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+    #[storage]
+    struct Storage {
+        public_key: felt252,
```
Comment (maciejka) [inline onchain/src/social_account.cairo:5]: `public_key` should be `u256`. See: https://github.com/keep-starknet-strange/joyboy/blob/4c572d060c946efb74b23f661a2576fe1cc1421f/onchain/src/bip340.cairo#L31C14-L31C18.

File: onchain/src/social_account.cairo [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        AccountCreated: AccountCreated,
```
Comment (maciejka) [inline onchain/src/social_account.cairo:11]: Is this finished?

File: onchain/src/social_account.cairo (hunk 2 above)
Comment (maciejka) [reply onchain/src/social_account.cairo:5]: Add a view method: `get_public_key`.

File: onchain/src/social_account.cairo [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+    }
+
+    #[constructor]
+    fn constructor(ref self: ContractState, public_key: public_key) {
+        self.public_key.write(public_key)
+    }
+//self.emit(AccountCreated { public_key: self.public_key });
```
Comment (maciejka) [inline onchain/src/social_account.cairo:18]: ?

File: onchain/src/social_account.cairo [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+}
```
Comment (maciejka) [inline onchain/src/social_account.cairo:20]: Add constructor test.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: onchain/src/social_account.cairo (hunk 3 above)
Comment (mubarak23) [reply onchain/src/social_account.cairo:11]: Yes, are we adding anything aside the event name

File: onchain/src/social_account.cairo (hunk 4 above)
Comment (mubarak23) [reply onchain/src/social_account.cairo:18]: i will should remove this line.

 when a constructor run when a contract is deployed, should we emit an event 

`AccountCreated`

**Comment by MubarakCinderbuild** [comment]
what could cause this error

```
Run scarb test
     Running test joyboy (snforge test)
snforge: command not found 
```
@maciejka

**Comment by mubarak23** [comment]
what could this error

```
Run scarb test
     Running test joyboy (snforge test)
snforge: command not found 
```

@maciejka

File: onchain/Scarb.toml [hunk 6]
```diff
@@ -6,3 +6,15 @@ edition = "2023_11"
 # See more keys and their definitions at https://docs.swmansion.com/scarb/docs/reference/manifest.html
 
 [dependencies]
+snforge_std = { git = "https://github.com/foundry-rs/starknet-foundry", tag = "v0.21.0" }
+openzeppelin = { git = "https://github.com/OpenZeppelin/cairo-contracts.git", tag = "v0.11.0" }
```
[the remark this answers is not in the document]
Comment (mubarak23) [reply onchain/Scarb.toml:10]: openzeppelin is not necessary, but we definitely need snforge_std in order for our test to run

File: onchain/src/social_account/social_account.cairo [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    }
+
+    #[derive(Drop, Serde)]
+    pub struct SocialAccountInitParams {
+        pub public_key: u256
+    }
+
+
+    #[constructor]
+    fn constructor(ref self: ContractState, init_params: SocialAccountInitParams) {
```
[the remark this answers is not in the document]
Comment (mubarak23) [reply onchain/src/social_account/social_account.cairo:31]: Alright

File: onchain/Scarb.lock [hunk 8]
```diff
@@ -4,3 +4,17 @@ version = 1
...
+ "snforge_std",
+]
+
+[[package]]
+name = "openzeppelin"
+version = "0.11.0"
+source = "git+https://github.com/OpenZeppelin/cairo-contracts.git?tag=v0.11.0#a83f36b23f1af6e160288962be4a2701c3ecbcda"
+
+[[package]]
+name = "snforge_std"
```
Comment (maciejka) [inline onchain/Scarb.lock:18]: snforge should be a dev dependency: https://foundry-rs.github.io/starknet-foundry/getting-started/first-steps.html.

**Comment by maciejka** [comment]
> what could cause this error
> 
> ```
> Run scarb test
>      Running test joyboy (snforge test)
> snforge: command not found 
> ```
> 
> @maciejka

snforge not installed correctly?

**Comment by maciejka** [comment]
> > what could cause this error
> > ```
> > Run scarb test
> >      Running test joyboy (snforge test)
> > snforge: command not found 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @maciejka
> 
> snforge not installed correctly?

You should add it to `.github/workflows`

**Comment by mubarak23** [comment]
> > > what could cause this error
> > > ```
> > > Run scarb test
> > >      Running test joyboy (snforge test)
> > > snforge: command not found 
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > @maciejka
> > 
> > 
> > snforge not installed correctly?
> 
> You should add it to `.github/workflows`

this is already setup inside  starknet-contracts.yml file of .github/workflows
```

  tests:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: software-mansion/setup-scarb@v1
        with:
          scarb-version: "nightly"
      - name: Run cairo tests
        run: scarb test
        working-directory: onchain
```
putting snforge_std as dev dependencies should fix this error

**Comment by mubarak23** [comment]
How do you install a new dependencies inside a github workflow

**Comment by maciejka** [comment]
> How do you install a new dependencies inside a github workflow

I suspect the problem is that `snforge` is not installed.

**Comment by mubarak23** [comment]
> > How do you install a new dependencies inside a github workflow
> 
> I suspect the problem is that `snforge` is not installed.

yes, from the error, snforge is not install

**Comment by mubarak23** [comment]
> > How do you install a new dependencies inside a github workflow
> 
> I suspect the problem is that `snforge` is not installed.

you will need to install it via github action

**Comment by maciejka** [comment]
> > > How do you install a new dependencies inside a github workflow
> > 
> > 
> > I suspect the problem is that `snforge` is not installed.
> 
> you will need to install it via github action

Add it to `.github/workflows/starknet-contracts.yml`

**Comment by mubarak23** [comment]
> > > > How do you install a new dependencies inside a github workflow
> > > 
> > > 
> > > I suspect the problem is that `snforge` is not installed.
> > 
> > 
> > you will need to install it via github action
> 
> Add it to `.github/workflows/starknet-contracts.yml`

i added this 

```
  test:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v4
      - uses: asdf-vm/actions/install@v3
      - run: snforge test
        working-directory: onchain

```

**Comment by mubarak23** [comment]
> > > > > How do you install a new dependencies inside a github workflow
> > > > 
> > > > 
> > > > I suspect the problem is that `snforge` is not installed.
> > > 
> > > 
> > > you will need to install it via github action
> > 
> > 
> > Add it to `.github/workflows/starknet-contracts.yml`
> 
> i added this
> 
> ```
>   test:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: actions/checkout@v4
>       - uses: asdf-vm/actions/install@v3
>       - run: snforge test
>         working-directory: onchain
> ```

change this to 


 ```
tests:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v3
      - uses: software-mansion/setup-scarb@v1
        with:
          scarb-version: 'nightly'
      - name: Run cairo tests
        uses: foundry-rs/setup-snfoundry@v3
        run: scarb test
        working-directory: onchain
```

File: onchain/src/social_account.cairo [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    };
+    use snforge_std as snf;
+    use snforge_std::{declare, ContractClassTrait};
+    use super::{ISocialPayAccountDispatcher, ISocialPayAccountDispatcherTrait};
+
+    const public_key: u256 = 45;
+
+    fn deploy_social_account() -> ContractAddress {
+        let contract = declare("SocialPayAccount");
+
```
Comment (maciejka) [inline onchain/src/social_account.cairo:56]: `AccountCreated` event emission should be verified here.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: onchain/src/social_account.cairo (hunk 9 above)
Comment (mubarak23) [reply onchain/src/social_account.cairo:56]: you mean, check that AccountCreated was actually emmited

Comment (mubarak23) [reply onchain/src/social_account.cairo:56]: how do check that event is actually emitted inside a function @maciejka

Comment (maciejka) [reply onchain/src/social_account.cairo:56]: With starknet foundry: https://foundry-rs.github.io/starknet-foundry/testing/testing-events.html

Comment (mubarak23) [reply onchain/src/social_account.cairo:56]: i have this 

    fn deploy_social_account() -> ContractAddress {
        let contract = declare("SocialPayAccount");

        let mut social_account_calldata = array![];
        public_key.serialize(ref social_account_calldata);

        let address = contract.precalculate_address(@social_account_calldata);
        
        let deployed_contract_address =  contract.deploy(@social_account_calldata).unwrap();
        
        let mut spy = spy_events(SpyOn::One(address));
        spy.assert_emitted(@array![
        (
            contract_address,
            AccountCreated { public_key }
            
        )
    ]);

       //  contract.deploy(@social_account_calldata).unwrap()
       deployed_contract_address
    }

**Comment by mubarak23** [comment]
@maciejka kindly review, thank you

File: onchain/src/social_account.cairo [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+        let address = contract.precalculate_address(@social_account_calldata);
+
+        let mut spy = spy_events(SpyOn::One(address));
+
+        let deployed_contract_address = contract.deploy(@social_account_calldata).unwrap();
+
+        spy.fetch_events();
+
+        assert(spy.events.len() == 1, 'there should be one event');
```
Comment (maciejka) [inline onchain/src/social_account.cairo:73]: Check that the emitted event public_key is correct.

**Review by maciejka — CHANGES_REQUESTED** [review]

File: onchain/src/social_account.cairo (hunk 10 above)
Comment (mubarak23) [reply onchain/src/social_account.cairo:73]: ok

Comment (mubarak23) [reply onchain/src/social_account.cairo:73]: Am getting this crazy error

`Trait has no implementation in context: core::traits::Into::<@core::array::Array::<core::felt252>, core::integer::u256>`

This is some part of the code
```

        let (_, event) = spy.events.at(0);
        assert(event.keys.at(0) == @selector!("AccountCreated"), 'Wrong event name');

        assert(public_key == event.data.into(), 'Public key is not the same');
```

And am importing

use traits::Into;  // this also throw this error -> Identifier not found.


I want both public_key and  event.data.into() to in the same type u256

Comment (mubarak23) [reply onchain/src/social_account.cairo:73]: from the documentation : https://foundry-rs.github.io/starknet-foundry/testing/testing-events.html

```
 let (_, event) = spy.events.at(0);
        assert(event.keys.at(0) == @selector!("AccountCreated"), 'Wrong event name');
```

we can only check the event name, we cannot check the event emitted values

Comment (maciejka) [reply onchain/src/social_account.cairo:73]: What about `assert_emitted`?

Comment (mubarak23) [reply onchain/src/social_account.cairo:75]: i did not really understand what they did here 

```
 spy.assert_emitted(@array![
        (
            contract_address,
            SpyEventsChecker::Event::FirstEvent(
                SpyEventsChecker::FirstEvent { some_data: 123 }
            )
        )
    ]);
```

Comment (mubarak23) [reply onchain/src/social_account.cairo:75]: does Into trait come with cairo core library by default

am getting the following error 

"Trait has no implementation in context: core::traits::Into::<@core::felt252, core::integer::u256>" 

```
for this code  let key_felt252 = event.keys.at(1);
        let key: u256 = key_felt252.into();
```

Comment (mubarak23) [reply onchain/src/social_account.cairo:76]: deferencing the *event.keys.at(1);  fix the above error

```
   let key_felt252 = *event.keys.at(1);
        let key: u256 = key_felt252.into();
        assert(key == public_key, 'Wrong Public Key');
```

**Review by maciejka — CHANGES_REQUESTED** [review]

**Comment by mubarak23** [comment]
@maciejka kindly review

File: onchain/src/social_account.cairo [hunk 11]
```diff
@@ -78,9 +76,9 @@ mod tests {
         let (_, event) = spy.events.at(0);
         assert(event.keys.at(0) == @selector!("AccountCreated"), 'Wrong event name');
 
-        let key_felt252 = *event.keys.at(1);
-        let key: u256 = key_felt252.into();
-        assert(key == public_key, 'Wrong Public Key');
+        let event_key: u256 = (*event.keys.at(1)).into();
```
Comment (maciejka) [inline onchain/src/social_account.cairo:79]: Do you really need the type declaration here?

**Review by maciejka — CHANGES_REQUESTED** [review]

File: onchain/src/social_account.cairo (hunk 11 above)
Comment (mubarak23) [reply onchain/src/social_account.cairo:79]: Yes, in order to type cast it as u256

Comment (mubarak23) [reply onchain/src/social_account.cairo:79]: but if this is clean i can change it 

  ` assert(public_key == (*event.keys.at(1)).into(), 'Wrong Public Key');`

Comment (mubarak23) [reply onchain/src/social_account.cairo]: this work         
```
let event_key = (*event.keys.at(1)).into();

assert(event_key == public_key, 'Wrong Public Key');
```

**Review by maciejka — APPROVED** [review]

**Pull request merged**